### PR TITLE
Split OCL; Eclipse Project Lookup Support

### PR DESCRIPTION
**Pull request by sdkrach**
- Split of the OCL Initialization into a separate bundle
- Added EclipseProjectScanner and EclipseProjectByClassRegistration
- Incremented Snapshot version, as builder interface is not compatible anymore

### CODE REVIEW

File: features/tools.mdsd.library.standalone.initialization.ocl.feature/feature.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<feature
+      id="tools.mdsd.library.standalone.initialization.ocl.feature"
+      label="OCL Standalone Initialization Feature"
+      version="0.1.0.qualifier">
```
Comment (seiferma) [inline features/tools.mdsd.library.standalone.initialization.ocl.feature/feature.xml]: We should use the license feature in all features but this can be a separate PR.

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectByClassRegistration.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * 
+ * The task looks for Eclipse Project files (".project") files and Jar-File Manifests
+ * ("META-INF/MANIFEST.MF"). It should be added to a Standalone Initialization sequence through
+ * {@link StandaloneInitializerBuilder#addCustomTask(InitializationTask)}.
+ * 
+ * @author Sebastian Krach
+ *
+ */
+public class EclipseProjectByClassRegistration implements InitializationTask {
+    private Class<?> clazz;
```
Comment (seiferma) [inline bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectByClassRegistration.java]: Could be final.

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        try {
+            Manifest manifest = new Manifest(Files.newInputStream(path, StandardOpenOption.READ));
+            var bundleName = manifest.getMainAttributes()
+                .getValue("Bundle-SymbolicName");
+            if (bundleName.indexOf(';') > 0) {
+                bundleName = bundleName.substring(0, bundleName.indexOf(';'));
+            }
+            return bundleName;
+        } catch (IOException e) {
+            throw new RuntimeException("Error reading the project file " + path.toString(), e);
```
Comment (seiferma) [inline bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java:128]: Why is there a need for RuntimeException instances? The signature of the task provides means for emitting a checked exception.

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        try {
+            document = DocumentBuilderFactory.newInstance()
+                .newDocumentBuilder()
+                .parse(Files.newInputStream(path, StandardOpenOption.READ));
+            return document.getDocumentElement()
+                .getElementsByTagName("name")
+                .item(0)
+                .getTextContent();
+        } catch (SAXException | IOException | ParserConfigurationException e) {
+            throw new RuntimeException("Error reading the project file " + path.toString(), e);
```
Comment (seiferma) [inline bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java:114]: Why is there a need for RuntimeException instances? The signature of the task provides means for emitting a checked exception.

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                    result = result.getParent();
+                }
+                return result.toFile();
+            }, (a, b) -> b)));
+
+        return res;
+    }
+
+    protected static String readProjectNameFromProjectFile(Path path) {
+        Document document;
```
Comment (seiferma) [inline bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java]: Could be moved into try block.

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                for (int i = 0; i < projectFilePath.getNameCount(); i++) {
+                    result = result.getParent();
+                }
+                return result.toFile();
+            }));
+
+        var manifestPath = Paths.get(JarFile.MANIFEST_NAME);
+        res.putAll(Files
+            .find(basePath, DEFAULT_MAX_DEPTH + manifestPath.getNameCount(),
+                    (path, attr) -> path.endsWith(Paths.get(JarFile.MANIFEST_NAME)))
```
Comment (seiferma) [inline bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java]: Reuse manifestPath here.

**Review by seiferma — CHANGES_REQUESTED** [review]
Looks good to me. Only the exception comment should be addressed or explained.

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java (hunk 3 above)
Comment (sdkrach) [reply bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java:128]: I really dislike checked exception, as they prevent you calling the method from within a lambda (which is done from within `Collectors.toMap`. I also do not see any benefit from using a checked exception here. Am I missing something?

File: features/tools.mdsd.library.standalone.initialization.ocl.feature/feature.xml (hunk 1 above)
Comment (sdkrach) [reply features/tools.mdsd.library.standalone.initialization.ocl.feature/feature.xml]: True

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectByClassRegistration.java (hunk 2 above)
Comment (sdkrach) [reply bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectByClassRegistration.java]: Done

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java (hunk 4 above)
Comment (sdkrach) [reply bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java:114]: Same as above

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java (hunk 6 above)
Comment (sdkrach) [reply bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java]: True

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectByClassRegistration.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        }
+    }
+
+    protected static Map<String, File> tryFindProjectContaining(Class<?> clz) throws StandaloneInitializationException {
+        Map<String, File> result = Collections.emptyMap();
+        try {
+            var classLocation = Paths.get(clz.getProtectionDomain()
+                .getCodeSource()
+                .getLocation()
+                .toURI());
```
Comment (seiferma) [inline bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectByClassRegistration.java:52]: see  #2

Comment (seiferma) [reply bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectByClassRegistration.java:52]: nvm this does not apply here.

Comment (sdkrach) [reply bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectByClassRegistration.java:52]: Does not apply, as Paths.get(URI) should handle encoded URIs well.

File: bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java (hunk 3 above)
Comment (seiferma) [reply bundles/tools.mdsd.library.standalone.initialization/src/tools/mdsd/library/standalone/initialization/core/EclipseProjectScanner.java:128]: The only downside I see is that callers are not aware of the fact that a runtime exception can be thrown as part of the regular execution logic. However, this does not seem an important case because callers usually cannot recover from an initialization issue. The runtime exception contains all information to fix the issue, so this should be fine.

**Review by seiferma — APPROVED** [review]
lgtm
